services/src/main/java/org/keycloak/protocol/saml/SamlService.java

```diff
@@ -1079,26 +1147,147 @@ private Document createArtifactResolve(String issuer, String artifact) throws Pr
+    }
+
+    private class ArtifactResolutionRunnable implements ScheduledTask{
+
+        private AsyncResponse asyncResponse;
+        private URI clientArtifactBindingURI;
+        private String relayState;
+        private Document doc;
+        private UriInfo uri;
+        private String realmId;
+        private HttpHeaders httpHeaders;
+        private ClientConnection connection;
+        private org.jboss.resteasy.spi.HttpResponse response;
+        private HttpRequest request;
+        private String bindingType;
+
+        public ArtifactResolutionRunnable(String bindingType, AsyncResponse asyncResponse, Document doc, URI clientArtifactBindingURI, String relayState){
+            this.asyncResponse = asyncResponse;
+            this.doc = doc;
+            this.clientArtifactBindingURI = clientArtifactBindingURI;
+            this.relayState = relayState;
+            this.uri = session.getContext().getUri();
+            this.realmId = realm.getId();
+            this.httpHeaders = new ResteasyHttpHeaders(headers.getRequestHeaders());
+            this.connection = new ClientConnectionWrapper();
+            this.response = Resteasy.getContextData(org.jboss.resteasy.spi.HttpResponse.class);
+            this.request = Resteasy.getContextData(HttpRequest.class);
+            this.bindingType = bindingType;
         }
 
-        byte[] source = new byte[20];
-        for (int i = 0; i < 20; i++) {
-            source[i] = artifact[i+4];
+
+        public void run(KeycloakSession session){
+            // Initialize context
+            Resteasy.pushContext(UriInfo.class, uri);
+
+            KeycloakTransaction tx = session.getTransactionManager();
+            Resteasy.pushContext(KeycloakTransaction.class, tx);
+
+            Resteasy.pushContext(KeycloakSession.class, session);
+            Resteasy.pushContext(HttpHeaders.class, httpHeaders);
+            Resteasy.pushContext(org.jboss.resteasy.spi.HttpResponse.class, response);
+            Resteasy.pushContext(HttpRequest.class, request);
+
+            session.getContext().setConnection(connection);
+            Resteasy.pushContext(ClientConnection.class, connection);
+
+            RealmManager realmManager = new RealmManager(session);
+            RealmModel realm = realmManager.getRealmByName(realmId);
+            if (realm == null) {
+                throw new NotFoundException("Realm does not exist");
+            }
+            session.getContext().setRealm(realm);
+
+            // Call Artifact Resolution Service
+            HttpResponse result;
+
+            try {
+                // TODO: If we use the HttpClient coming from Provider, some tests fails when they are run all together.
+                // HttpClientProvider httpClientProvider = session.getProvider(HttpClientProvider.class);
+                // HttpClient httpClient = httpClientProvider.getHttpClient();
+                CloseableHttpClient httpClient = HttpClients.createDefault();
```

Comment:
Client does not to be closed... Shouldn't it be moved in the try statement?
```
try (CloseableHttpClient httpClient=HttpClients.createDefault()) {
}
```

Reply:
Watch out, if you close the client before processing the "result", you might have issues.

Reply:
This is something that Hynek will have to fix anyway....